scripts/check_integrity.py

```diff
@@ -0,0 +1,251 @@
+database_file = '../literature.bib'
+
+title_regex = re.compile(r"^\{.*\}$", re.DOTALL)
+pages_regex = re.compile(r"^[a-zA-Z0-9,.:()]+(--[a-zA-Z0-9,.:()]+)?$", re.DOTALL)
+author_regex = re.compile(r"([^,]+),[^,]+(\s+and\s+([^,]+),[^,]+)*", re.DOTALL)
+
+
+def get_path(file_name):
+    return os.path.join(script_dir, file_name)
+
+
+## Checks if parameter is set in config file
+def is_set(key):
+    if not key in settings:
+        print('WARNING settings ' + key + ' is not defined!')
+        return False
+    return settings[key]
+
+
+## Adds a problem to the list of problems of a given entry
+def add_problem(entry, problem_type, problem_message):
+    entry_id = entry['ID']
+    if entry_id not in problems:
+        problems[entry_id] = []
+    problems[entry_id].append((problem_type, problem_message))
+
+
+## Checks if fields declared as string fields contain string values
+def check_string_fields(entry):
+    for field in string_fields:
+        if is_set('string_' + field):
+            if field in entry and isinstance(entry[field], str):
+                add_problem(entry, 'non_string_' + field, entry[field])
+
+
+## Creates bibtex strings for fields that are declared as string fields but do not contains string values
+def gen_string_fields(entry):
+    for field in string_fields:
+        if field in entry and isinstance(entry[field], str):
+            field_value = entry[field]
+            key = re.sub('[^A-Z]', '', field_value)
+            new_strings[field_value] = (field, key, field_value)
+
+
+## Checks that required, wanted or optional fields exists for a given entry or is declared as missing
+def check_field_type_exists(entry, field_type):
+    if is_set(field_type + '_fields'):
+        entry_type=entry['ENTRYTYPE']
+        if entry_type in fields:
+            required_fields = (fields[entry_type])[field_type] or []
+        else:
+            required_fields = []
+        entry_id = entry['ID']
+        entry_fields = set(entry.keys())
+        if 'missing' in entry:
+            missing = [x for x in entry['missing'].get_value().split(',') if x]
+        else:
+            missing = []
+        for field in required_fields:
+            if isinstance(field, list):
+                present = False
+                for alternative_field in field:
+                    if alternative_field in entry_fields or alternative_field in missing:
+                        present = True
+                        break
+                if not present:
+                    add_problem(entry, 'missing_' + field_type + '_field', field)
+            else:
+                if field not in missing and field not in entry_fields:
+                    add_problem(entry, 'missing_' + field_type + '_field', field)
+
+
+def check_field_types_exists(entry):
+    check_field_type_exists(entry, 'required')
+    check_field_type_exists(entry, 'wanted')
+    check_field_type_exists(entry, 'optional')
```

Comment:
Can you document these types and their meaning? Would be nice to have them as constants.

Reply:
done